test/unit/shim/promise-shim.test.js

```diff
@@ -237,44 +236,38 @@ tap.test('PromiseShim', (t) => {
       })
     })
 
-    t.test('should reinstate lost context', (t) => {
-      helper.runInTransaction(agent, (tx) => {
+    t.test('should reinstate lost context', async (t) => {
+      t.autoend()
+      helper.runInTransaction(agent, async (tx) => {
         shim.setClass(TestPromise)
         const WrappedPromise = shim.wrapConstructor(TestPromise)
 
         // Wrapping then is required to make sure the then callback is wrapped
         // with context propagation.
         shim.wrapThen(TestPromise.prototype, 'then')
 
-        asyncLib.series(
-          [
-            (cb) => {
-              t.sameTransaction(agent.getTransaction(), tx)
-              new WrappedPromise((resolve) => {
-                t.sameTransaction(agent.getTransaction(), tx)
-                resolve() // <-- Resolve will lose context.
-              })
-                .then(() => {
-                  t.sameTransaction(agent.getTransaction(), tx)
-                  cb()
-                })
-                .catch(cb)
-            },
-            (cb) => {
-              t.sameTransaction(agent.getTransaction(), tx)
-              new WrappedPromise((resolve) => {
-                t.sameTransaction(agent.getTransaction(), tx)
-                helper.runOutOfContext(resolve) // <-- Context loss before resolve.
-              })
-                .then(() => {
-                  t.sameTransaction(agent.getTransaction(), tx)
-                  cb()
-                })
-                .catch(cb)
+        const txTest = async (runOutOfContext, runNext) => {
```

Comment:
I tried really hard to rewrite this in a way that reads a bit easier, but I ran into the exact same problems that you described in standup 😭 

The problem is definitely that the second promise never resolves, so maybe something with that `runOutOfContext` helper is deeply tied to the `async.series` execution? Gonna do some more digging because this is really hard to grok and making sure that context restoration works is important

Reply:
I did dig into the `runOutOfContext` helper, and I've come around to thinking that it's not an interaction between this and `async`, particularly. The `async` library is used just to slow this down and avoid concurrent executions. We'd be able to solve that by making these functions truly async functions, but for me that would have required an extensive refactoring of this entire test. Which might be better--it'd at least be more clear. 

I'm not overly fond of how I solved this problem, recursing through. It removes the abstraction through `async`, at least, but the recursion isn't easy to follow.

Reply:
Yeah, getting rid of async once and for all is enough of a win that i'm willing to eat the strangeness of the test for now

Reply:
If nothing else, at least the full function code is available here, instead of depending on code deeply buried in that library.